tests/prague/eip7692_eof_v1/eip3540_eof_v1/test_migrated_valid_invalid.py

```diff
@@ -0,0 +1,288 @@
+"""
+
+import pytest
+
+from ethereum_test_tools import EOFTestFiller
+from ethereum_test_tools import Opcodes as Op
+from ethereum_test_tools.eof.v1 import Container, EOFException, Section
+from ethereum_test_tools.eof.v1.constants import NON_RETURNING_SECTION
+
+from .. import EOF_FORK_NAME
+
+REFERENCE_SPEC_GIT_PATH = "EIPS/eip-3540.md"
+REFERENCE_SPEC_VERSION = "8dcb0a8c1c0102c87224308028632cc986a61183"
+
+pytestmark = pytest.mark.valid_from(EOF_FORK_NAME)
+
+
+@pytest.mark.parametrize(
+    "eof_code,exception",
+    [
+        pytest.param(
+            # Deployed code without data section
+            Container(
+                name="EOF1V3540_0001",
+                sections=[
+                    Section.Code(code=Op.PUSH1[0] + Op.POP + Op.STOP, max_stack_height=1),
+                ],
+            ),
+            None,
+            id="EOF1V3540_0001",
+        ),
+        pytest.param(
+            # Deployed code with data section
+            Container(
+                name="EOF1V3540_0002",
+                sections=[
+                    Section.Code(code=Op.PUSH1[0] + Op.POP + Op.STOP, max_stack_height=1),
+                    Section.Data("aabbccdd"),
+                ],
+            ),
+            None,
+            id="EOF1V3540_0002",
+        ),
+        pytest.param(
+            # No data section contents (valid according to relaxed validation)
+            Container(
+                name="EOF1V3540_0003",
+                sections=[
+                    Section.Code(code=Op.INVALID),
+                    Section.Data(custom_size=2),
+                ],
+            ),
+            None,
+            id="EOF1V3540_0003",
+        ),
+        pytest.param(
+            # Data section contents incomplete (valid according to relaxed validation)
+            Container(
+                name="EOF1V3540_0004",
+                sections=[
+                    Section.Code(code=Op.INVALID),
+                    Section.Data("aa", custom_size=2),
+                ],
+            ),
+            None,
+            id="EOF1V3540_0004",
+        ),
+        pytest.param(
+            # Type section size incomplete
+            bytes.fromhex("ef00010100"),
+            EOFException.INCOMPLETE_SECTION_SIZE,
+            id="EOF1I3540_0011",
+        ),
+        pytest.param(
+            # Empty code section with non-empty data section
+            bytes.fromhex("ef000101000402000100000400020000000000aabb"),
+            EOFException.ZERO_SECTION_SIZE,
+            id="EOF1I3540_0012",
+        ),
+        pytest.param(
+            # No data section after code section size
+            bytes.fromhex("ef00010100040200010001"),
+            EOFException.MISSING_HEADERS_TERMINATOR,
+            id="EOF1I3540_0017",
+        ),
+        pytest.param(
+            # No data size
+            bytes.fromhex("ef0001010004020001000104"),
+            EOFException.MISSING_HEADERS_TERMINATOR,
```

Comment:
We can remove `"no_data_section_size"` from `tests/prague/eip7692_eof_v1/eip3540_eof_v1/container.py` (which is actually incorrect) in this same PR because this test supersedes it.